### ISSUE TITLE
fatal error: concurrent map iteration and map write

### ISSUE THREAD

**Issue by BuriedStPatrick** [CrunchyData/pg_tileserv#108]
We run pg_tileserv in an Azure AKS Kubernetes cluster and are experiencing some crashes from time to time, seemingly when under load as far as I can tell. My suspicion is that it's something to do with our backend being too slow to respond which results in this crash, but I'm not entirely sure.

In any case, it seems to cause a quite low-level concurrency-exception which then crashes the pod completely. We're getting around the issue by simply scaling the replicas to handle failover in case one goes down. Plus, we've added a cache in front to reduce the load.

In case this is caused by a slow backend, I would expect it to just take a bit longer to respond, so in case I'm wrong in that assumption, I hope someone can point me in the right direction :)

Image: `pramsey/pg_tileserv:20210421`
Kubernetes: `1.21.2`
Node image version: `AKSUbuntu-1804gen2containerd-2021.07.17`
Backend:
* Azure Database for PostgreSQL server
* Memory Optimized, 2v Cores
* Postgres : `11`

Logs:
```
time="2021-08-20T07:09:54Z" level=info msg="pg_tileserv latest"
time="2021-08-20T07:09:54Z" level=info msg="Run with --help parameter for commandline options"
time="2021-08-20T07:09:54Z" level=info msg="Using database connection info from environment variable DATABASE_URL"
time="2021-08-20T07:09:54Z" level=info msg="Using config file: /config/pg_tileserv.toml"
time="2021-08-20T07:09:54Z" level=info msg="Serving HTTP  at http://0.0.0.0:7800/"
time="2021-08-20T07:09:54Z" level=info msg="Serving HTTPS at http://0.0.0.0:7801/"
time="2021-08-20T07:09:54Z" level=info msg="Using CoordinateSystem.SRID 3857 with bounds [-2.00375083427892e+07, -2.00375083427892e+07, 2.00375083427892e+07, 2.00375083427892e+07]"
time="2021-08-20T07:09:55Z" level=info msg="Connected as '<REDACTED>' to '<REDACTED>' @ '<REDACTED>'"
time="2021-08-20T07:09:55Z" level=info msg="Serving HTTP  at 0.0.0.0:7800"
time="2021-08-20T07:10:41Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:10:51Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:11:01Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:11:01Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:11:06Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:11:11Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:11:11Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:11:21Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:11:21Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:11:31Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:11:31Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:11:41Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:11:41Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:11:51Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:11:51Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:12:01Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:12:01Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:12:11Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:12:11Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:12:21Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:12:21Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:12:31Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:12:31Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:12:41Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:12:41Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:12:51Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:12:51Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:13:01Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:13:01Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:13:11Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:13:11Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:13:16Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:13:21Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:13:21Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:13:31Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:13:31Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:13:41Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:13:41Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:13:51Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:13:51Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:14:01Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:14:01Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:14:11Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:14:11Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:14:21Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:14:21Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:14:31Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:14:31Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:14:41Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:14:41Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:14:51Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:14:51Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:15:01Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:15:01Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:15:11Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:15:11Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:15:21Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:15:21Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:15:26Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:15:31Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:15:31Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:15:41Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:15:41Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:15:51Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:15:51Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:16:01Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:16:01Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:16:11Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:16:11Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:16:21Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:16:21Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:16:31Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:16:31Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:16:41Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:16:41Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:16:51Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:16:51Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:17:01Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:17:01Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:17:11Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:17:11Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:17:21Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:17:21Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:17:31Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:17:31Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:17:36Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:17:41Z" level=info msg="GET /" method=GET url=/
time="2021-08-20T07:17:41Z" level=info msg="GET /" method=GET url=/
fatal error: concurrent map iteration and map write

goroutine 282 [running]:
runtime.throw(0x9cc01e, 0x1)
	/home/travis/.gimme/versions/go/src/runtime/panic.go:1198 +0x71 fp=0xc000275008 sp=0xc000274fd8 pc=0x435231
runtime.mapiternext(0x40f765)
	/home/travis/.gimme/versions/go/src/runtime/map.go:858 +0x4eb fp=0xc000275078 sp=0xc000275008 pc=0x4112ab
runtime.mapiterinit(0xc000275310, 0xc0001c10d0, 0x0)
	/home/travis/.gimme/versions/go/src/runtime/map.go:848 +0x236 fp=0xc0002750a0 sp=0xc000275078 pc=0x410d76
main.getJSONLayers(0xc0000d45b0)
	/home/travis/gopath/src/github.com/CrunchyData/pg_tileserv/layer.go:99 +0x95 fp=0xc000275258 sp=0xc0002750a0 pc=0x8aac15
main.requestListHTML(0xa73538, 0xc0002ea0e0, 0x9b4d89)
	/home/travis/gopath/src/github.com/CrunchyData/pg_tileserv/main.go:253 +0x1b3 fp=0xc000275430 sp=0xc000275258 pc=0x8b27d3
main.tileAppHandler.ServeHTTP(0x9e56d0, 0xa73538, 0xc0002ea0e0, 0xc00036e300)
	/home/travis/gopath/src/github.com/CrunchyData/pg_tileserv/main.go:380 +0x242 fp=0xc0002756a0 sp=0xc000275430 pc=0x8b3c22
github.com/gorilla/mux.(*Router).ServeHTTP(0xc0003969c0, 0xa73538, 0xc0002ea0e0, 0xc00036e000)
	/home/travis/gopath/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x1e2 fp=0xc0002757d0 sp=0xc0002756a0 pc=0x870462
github.com/gorilla/handlers.(*cors).ServeHTTP(0xc00039b0e0, 0xa73538, 0xc0002ea0e0, 0xc00036e000)
	/home/travis/gopath/pkg/mod/github.com/gorilla/handlers@v1.5.1/cors.go:54 +0x3fa fp=0xc000275918 sp=0xc0002757d0 pc=0x86e71a
github.com/gorilla/handlers.CompressHandlerLevel.func1(0xa73538, 0xc0002ea0e0, 0xc00036e000)
	/home/travis/gopath/pkg/mod/github.com/gorilla/handlers@v1.5.1/compress.go:100 +0x6ee fp=0xc000275a68 sp=0xc000275918 pc=0x86dfee
net/http.HandlerFunc.ServeHTTP(0xc00017da68, 0xa73538, 0xc0002ea0e0, 0x10)
	/home/travis/.gimme/versions/go/src/net/http/server.go:2066 +0x2f fp=0xc000275a90 sp=0xc000275a68 pc=0x7aeb8f
main.setCacheControl.func1(0xa73538, 0xc0002ea0e0, 0xc000020030)
	/home/travis/gopath/src/github.com/CrunchyData/pg_tileserv/main.go:411 +0x1e6 fp=0xc000275b20 sp=0xc000275a90 pc=0x8b4226
net/http.HandlerFunc.ServeHTTP(0xc000168000, 0xa73538, 0xc0002ea0e0, 0xc00017db50)
	/home/travis/.gimme/versions/go/src/net/http/server.go:2066 +0x2f fp=0xc000275b48 sp=0xc000275b20 pc=0x7aeb8f
net/http.serverHandler.ServeHTTP(0xc0004f40f0, 0xa73538, 0xc0002ea0e0, 0x6)
	/home/travis/.gimme/versions/go/src/net/http/server.go:2886 +0x8e fp=0xc000275b78 sp=0xc000275b48 pc=0x7b1d8e
net/http.(*conn).serve(0xc00014e640, 0xa75d60, 0xc00050e0c0)
	/home/travis/.gimme/versions/go/src/net/http/server.go:1949 +0xb08 fp=0xc000275fb8 sp=0xc000275b78 pc=0x7adc68
net/http.(*Server).Serve·dwrap·82()
	/home/travis/.gimme/versions/go/src/net/http/server.go:3012 +0x2e fp=0xc000275fe0 sp=0xc000275fb8 pc=0x7b260e
runtime.goexit()
	/home/travis/.gimme/versions/go/src/runtime/asm_amd64.s:1583 +0x1 fp=0xc000275fe8 sp=0xc000275fe0 pc=0x464741
created by net/http.(*Server).Serve
	/home/travis/.gimme/versions/go/src/net/http/server.go:3012 +0x4e5

goroutine 1 [chan receive, 6 minutes]:
main.handleRequests()
	/home/travis/gopath/src/github.com/CrunchyData/pg_tileserv/main.go:510 +0x879
main.main()
	/home/travis/gopath/src/github.com/CrunchyData/pg_tileserv/main.go:194 +0xe79

goroutine 19 [select, 1 minutes]:
github.com/jackc/pgx/v4/pgxpool.(*Pool).backgroundHealthCheck(0xc0000daaf0)
	/home/travis/gopath/pkg/mod/github.com/jackc/pgx/v4@v4.9.2/pgxpool/pool.go:322 +0x90
created by github.com/jackc/pgx/v4/pgxpool.ConnectConfig
	/home/travis/gopath/pkg/mod/github.com/jackc/pgx/v4@v4.9.2/pgxpool/pool.go:205 +0x2e5

goroutine 44 [IO wait]:
internal/poll.runtime_pollWait(0x800000, 0x72)
	/home/travis/.gimme/versions/go/src/runtime/netpoll.go:229 +0x89
internal/poll.(*pollDesc).wait(0xc0000d3580, 0xc000348000, 0xc00003b800)
	/home/travis/.gimme/versions/go/src/internal/poll/fd_poll_runtime.go:84 +0x32
internal/poll.(*pollDesc).waitRead(...)
	/home/travis/.gimme/versions/go/src/internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Accept(0xc000228080)
	/home/travis/.gimme/versions/go/src/internal/poll/fd_unix.go:402 +0x225
net.(*netFD).accept(0xc0000d3580)
	/home/travis/.gimme/versions/go/src/net/fd_unix.go:173 +0x35
net.(*TCPListener).accept(0xc0000a8ee8)
	/home/travis/.gimme/versions/go/src/net/tcpsock_posix.go:140 +0x28
net.(*TCPListener).Accept(0x7b2385)
	/home/travis/.gimme/versions/go/src/net/tcpsock.go:262 +0x3d
net/http.(*Server).Serve(0x9b4537, 0xa73358, 0xc0000a8ee8)
	/home/travis/.gimme/versions/go/src/net/http/server.go:2980 +0x394
net/http.(*Server).ListenAndServe(0x0)
	/home/travis/.gimme/versions/go/src/net/http/server.go:2909 +0x7d
main.handleRequests.func1()
	/home/travis/gopath/src/github.com/CrunchyData/pg_tileserv/main.go:469 +0x25
created by main.handleRequests
	/home/travis/gopath/src/github.com/CrunchyData/pg_tileserv/main.go:466 +0x2e5

goroutine 283 [IO wait]:
internal/poll.runtime_pollWait(0x0, 0x72)
	/home/travis/.gimme/versions/go/src/runtime/netpoll.go:229 +0x89
internal/poll.(*pollDesc).wait(0xc00033a080, 0xc0004f4101, 0x414fc5)
	/home/travis/.gimme/versions/go/src/internal/poll/fd_poll_runtime.go:84 +0x32
internal/poll.(*pollDesc).waitRead(...)
	/home/travis/.gimme/versions/go/src/internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Read(0x0, 0xc0004f4101, 0x1, 0x1)
	/home/travis/.gimme/versions/go/src/internal/poll/fd_unix.go:167 +0x259
net.(*netFD).Read(0xc00033a080, 0xc0004f4101, 0x406e1d, 0xc0004e2c00)
	/home/travis/.gimme/versions/go/src/net/fd_posix.go:56 +0x29
net.(*conn).Read(0xc00034df60, 0xc0004f4101, 0x995c20, 0xc00040c0e0)
	/home/travis/.gimme/versions/go/src/net/net.go:183 +0x45
net/http.(*connReader).backgroundRead(0xc00044ac00)
	/home/travis/.gimme/versions/go/src/net/http/server.go:692 +0x3f
created by net/http.(*connReader).startBackgroundRead
	/home/travis/.gimme/versions/go/src/net/http/server.go:688 +0xcf

goroutine 46 [syscall, 6 minutes]:
os/signal.signal_recv()
	/home/travis/.gimme/versions/go/src/runtime/sigqueue.go:169 +0x98
os/signal.loop()
	/home/travis/.gimme/versions/go/src/os/signal/signal_unix.go:24 +0x19
created by os/signal.Notify.func1.1
	/home/travis/.gimme/versions/go/src/os/signal/signal.go:151 +0x2c

goroutine 281 [runnable]:
github.com/spf13/cast.ToStringE(0x8fa9e0, 0xa5c2b0)
	/home/travis/gopath/pkg/mod/github.com/spf13/cast@v1.3.1/caste.go:799 +0x5db
github.com/spf13/cast.ToString(...)
	/home/travis/gopath/pkg/mod/github.com/spf13/cast@v1.3.1/cast.go:103
github.com/spf13/viper.(*Viper).GetString(0x203000, 0x9b70c4, 0x30)
	/home/travis/gopath/pkg/mod/github.com/spf13/viper@v1.7.1/viper.go:793 +0x25
github.com/spf13/viper.GetString(...)
	/home/travis/gopath/pkg/mod/github.com/spf13/viper@v1.7.1/viper.go:791
main.serverURLHost(0xc000442800)
	/home/travis/gopath/src/github.com/CrunchyData/pg_tileserv/util.go:62 +0x3b
main.serverURLBase(0xc000271310)
	/home/travis/gopath/src/github.com/CrunchyData/pg_tileserv/util.go:47 +0x19
main.getJSONLayers(0xc0000d45b0)
	/home/travis/gopath/src/github.com/CrunchyData/pg_tileserv/layer.go:98 +0x4b
main.requestListHTML(0xa73538, 0xc0003940e0, 0x9b4d89)
	/home/travis/gopath/src/github.com/CrunchyData/pg_tileserv/main.go:253 +0x1b3
main.tileAppHandler.ServeHTTP(0x9e56d0, 0xa73538, 0xc0003940e0, 0xc000442800)
	/home/travis/gopath/src/github.com/CrunchyData/pg_tileserv/main.go:380 +0x242
github.com/gorilla/mux.(*Router).ServeHTTP(0xc0003969c0, 0xa73538, 0xc0003940e0, 0xc000442500)
	/home/travis/gopath/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x1e2
github.com/gorilla/handlers.(*cors).ServeHTTP(0xc00039b0e0, 0xa73538, 0xc0003940e0, 0xc000442500)
	/home/travis/gopath/pkg/mod/github.com/gorilla/handlers@v1.5.1/cors.go:54 +0x3fa
github.com/gorilla/handlers.CompressHandlerLevel.func1(0xa73538, 0xc0003940e0, 0xc000442500)
	/home/travis/gopath/pkg/mod/github.com/gorilla/handlers@v1.5.1/compress.go:100 +0x6ee
net/http.HandlerFunc.ServeHTTP(0xc000053a68, 0xa73538, 0xc0003940e0, 0x10)
	/home/travis/.gimme/versions/go/src/net/http/server.go:2066 +0x2f
main.setCacheControl.func1(0xa73538, 0xc0003940e0, 0xc000408750)
	/home/travis/gopath/src/github.com/CrunchyData/pg_tileserv/main.go:411 +0x1e6
net/http.HandlerFunc.ServeHTTP(0xc000109080, 0xa73538, 0xc0003940e0, 0xc000053b50)
	/home/travis/.gimme/versions/go/src/net/http/server.go:2066 +0x2f
net/http.serverHandler.ServeHTTP(0xc00049c930, 0xa73538, 0xc0003940e0, 0x6)
	/home/travis/.gimme/versions/go/src/net/http/server.go:2886 +0x8e
net/http.(*conn).serve(0xc00014e460, 0xa75d60, 0xc000324d00)
	/home/travis/.gimme/versions/go/src/net/http/server.go:1949 +0xb08
created by net/http.(*Server).Serve
	/home/travis/.gimme/versions/go/src/net/http/server.go:3012 +0x4e5

goroutine 315 [IO wait]:
internal/poll.runtime_pollWait(0x0, 0x72)
	/home/travis/.gimme/versions/go/src/runtime/netpoll.go:229 +0x89
internal/poll.(*pollDesc).wait(0xc00033a000, 0xc00049c941, 0x4)
	/home/travis/.gimme/versions/go/src/internal/poll/fd_poll_runtime.go:84 +0x32
internal/poll.(*pollDesc).waitRead(...)
	/home/travis/.gimme/versions/go/src/internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Read(0x0, 0xc00049c941, 0x1, 0x1)
	/home/travis/.gimme/versions/go/src/internal/poll/fd_unix.go:167 +0x259
net.(*netFD).Read(0xc00033a000, 0xc00049c941, 0x40c8bd, 0xc000040740)
	/home/travis/.gimme/versions/go/src/net/fd_posix.go:56 +0x29
net.(*conn).Read(0xc000040760, 0xc00049c941, 0xc000324050, 0xc000204c80)
	/home/travis/.gimme/versions/go/src/net/net.go:183 +0x45
net/http.(*connReader).backgroundRead(0xc00033be00)
	/home/travis/.gimme/versions/go/src/net/http/server.go:692 +0x3f
created by net/http.(*connReader).startBackgroundRead
	/home/travis/.gimme/versions/go/src/net/http/server.go:688 +0xcf
```

**Comment by pramsey** [comment]
We did have some concurrency issues around a shared map that were addressed moderately recently, are you using a fresh version?

**Comment by BuriedStPatrick** [comment]
Thanks for getting back! Yes, I took a look at the latest release notes and it did sound an awful lot like our issue should have been fixed in the latest release 1.0.8.

We run the version tag `20210421` which, from what I can tell on Docker hub, seems to be the same 
digest as the `latest` tag. Perhaps the image isn't up to date on Docker hub?

**Comment by adamtimm** [comment]
> We run pg_tileserv in an Azure AKS Kubernetes cluster and are experiencing some crashes from time to time, seemingly when under load as far as I can tell. My suspicion is that it's something to do with our backend being too slow to respond which results in this crash, but I'm not entirely sure.
> 

If you're looking for a better performing alternative to Postgres on Azure AKS check out https://www.crunchydata.com/products/crunchy-bridge/ which includes full support for your Postgres cluster and pg_tileserv/pg_featureserv.
